src/ORM/Rule/ValidCount.php

```diff
@@ -0,0 +1,70 @@
+<?php
+/**
+ * CakePHP(tm) : Rapid Development Framework (http://cakephp.org)
+ * Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ *
+ * Licensed under The MIT License
+ * For full copyright and license information, please see the LICENSE.txt
+ * Redistributions of files must retain the above copyright notice.
+ *
+ * @copyright     Copyright (c) Cake Software Foundation, Inc. (http://cakefoundation.org)
+ * @link          http://cakephp.org CakePHP(tm) Project
+ * @since         3.0.0
+ * @license       http://www.opensource.org/licenses/mit-license.php MIT License
+ */
+namespace Cake\ORM\Rule;
+
+use Cake\Datasource\EntityInterface;
+use Cake\ORM\Association;
+use Cake\Validation\Validation;
+use RuntimeException;
+
+/**
+ * Validates the count of associated records.
+ */
+class ValidCount
+{
+
+    /**
+     * The field to check
+     *
+     * @var array
+     */
+    protected $_field;
+
+    /**
+     * Constructor.
+     *
+     * @param string $field The field to check the count on.
+     */
+    public function __construct($field)
+    {
+        $this->_field = $field;
+    }
+
+    /**
+     * Performs the count check
+     *
+     * @param \Cake\Datasource\EntityInterface $entity The entity from where to extract the fields
+     * @param array $options Options passed to the check.
+     * @return bool
+     */
+    public function __invoke(EntityInterface $entity, array $options)
+    {
+        $value = $entity->{$this->_field};
+        if (!is_array($value)) {
+            return false;
+        }
+
+        if (isset($value['_ids'])) {
+            if (!is_array($value['_ids'])) {
+                return false;
+            }
+            $count = count($value['_ids']);
+        } else {
+            $count = count($value);
+        }
+
+        return Validation::comparison($count, $options['operator'], $options['count']);
+    }
```

Comment:
@lorenzo OK, I'll be out until Sunday late night, so I'll finish it Sunday night or next week. I'll do the DB query count in another as soon as I find the time to do so.

But another question. When I do the DB count, I'll need to compare the DB count with the entities (?) I'm getting instead of the ids and check for the count of new entities vs existing and then calc the count based on the total of existing + new. Correct? Or am I missing something?